### PR TITLE
Fix | Fix for ResultSets being consumed when reading warnings

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Fixes issue #969

### CODE REVIEW

**Comment by cheenamalhotra** [comment]
Jars for testing: [mssql-jdbc-PR991.zip](https://github.com/Microsoft/mssql-jdbc/files/2969011/mssql-jdbc-PR991.zip)

**Comment by biasb** [comment]
I have verified that the jre8 version jar works in our test suite.

**Comment by cheenamalhotra** [comment]
Hi @biasb 

I've added the scenario as you mentioned, please let me know if I misunderstood anything.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java [hunk 1]
```diff
@@ -257,7 +257,7 @@ boolean onDone(TDSReader tdsReader) throws SQLServerException {
         }
 
         ExecDoneHandler execDoneHandler = new ExecDoneHandler();
-        TDSParser.parse(resultsReader(), execDoneHandler);
+        TDSParser.parse(resultsReader(), execDoneHandler, false);
```
Comment (sehrope) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java]: Why explicitly add `false` parameter here? The two arg version of `TDSParser.parse(...)` has been updated to default to false and call the three arg version via `parse(..., ..., false)` so its not necessary. Ditto for the other ten places that have been updated with the new `false` parameter. If it's meant to be explicit everywhere then the old method should be removed and it should be added across the board but there are still other two-arg invocations.

File: src/main/java/com/microsoft/sqlserver/jdbc/tdsparser.java [hunk 2]
```diff
@@ -23,10 +23,11 @@
      * token handler says to stop by returning false from one of the token handling methods.
      */
     static void parse(TDSReader tdsReader, String logContext) throws SQLServerException {
-        parse(tdsReader, new TDSTokenHandler(logContext));
+        parse(tdsReader, new TDSTokenHandler(logContext), false);
     }
 
-    static void parse(TDSReader tdsReader, TDSTokenHandler tdsTokenHandler) throws SQLServerException {
+    static void parse(TDSReader tdsReader, TDSTokenHandler tdsTokenHandler,
+            boolean readOnlyWarningsFlag) throws SQLServerException {
```
Comment (sehrope) [inline src/main/java/com/microsoft/sqlserver/jdbc/tdsparser.java]: The javadoc for `parse(...)` should be updated to have the new parameter documented here. Took a bit of skimming through the code to understand what it does.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java (hunk 1 above)
Comment (cheenamalhotra) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java]: Hi @sehrope 

The existing 2 param parse call accepts String argument as second param to define `logContext` while the second method accepts `TokenHandler` instance. The new param is added only for specific case where only TDS_MSG tokens are to be read. I overloaded the original `parse` method so the driver calls will not be modified to include third parameter except where its truly desired.

File: src/main/java/com/microsoft/sqlserver/jdbc/tdsparser.java (hunk 2 above)
Comment (cheenamalhotra) [reply src/main/java/com/microsoft/sqlserver/jdbc/tdsparser.java]: Added.

File: src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java [hunk 3]
```diff
@@ -510,4 +511,89 @@ public void testResultSetMethods() throws SQLException {
             }
         }
     }
+
+    @Test
+    public void testMultipleResultSets() throws SQLException {
+        try (Statement stmt = connection.createStatement()) {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: Please create a new connection for the test.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerStatement.java [hunk 4]
```diff
@@ -1385,6 +1385,7 @@ final void clearLastResult() {
      * @return true if another result (ResultSet or update count) was available; false if there were no more results.
      */
     final boolean getNextResult(boolean clearFlag) throws SQLServerException {
+
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerStatement.java]: Please remove this.

File: src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java [hunk 5]
```diff
@@ -510,4 +511,89 @@ public void testResultSetMethods() throws SQLException {
...
+
+            boolean results = stmt.execute(SQL);
+            int rsCount = 0;
+
+            // Loop through the available result sets.
+            while (results) {
+                try (ResultSet rs = stmt.getResultSet()) {
+                    rsCount++;
+                    String firstColumnName = rs.getMetaData().getColumnName(1);
+                    switch (rsCount) {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: This `switch` statement seems to be redundant, `firstColumnValue` can be fetched using the column names instead in the second `switch statement`, which will also validate the column names at the same time (if column names do not match, exception will be thrown). 

```
            // Loop through the available result sets.
            while (results) {
                try (ResultSet rs = stmt.getResultSet()) {
                    rsCount++;
                    int i = 1;
                    while (rs.next()) {
                        String firstColumnValue = null;rs.getString(1);
                        switch (rsCount) {
                            case 1:
                                firstColumnValue = rs.getString("Name");
                                assert (firstColumnValue.equals(tableName));
                                break;
                            case 2:
                                firstColumnValue = rs.getString("Column_name");
                                assert (firstColumnValue.equalsIgnoreCase("c" + i++));
                                break;
                            case 3:
                                firstColumnValue = rs.getString("Identity");
                                assert (firstColumnValue.equalsIgnoreCase("c1"));
                                break;
                            case 4:
                                firstColumnValue = rs.getString("RowGuidCol");
                                assert (firstColumnValue.equalsIgnoreCase("No rowguidcol column defined."));
                                break;
                            case 5:
                                firstColumnValue = rs.getString("Data_located_on_filegroup");
                                assert (firstColumnValue.equalsIgnoreCase("PRIMARY"));
                                break;
                        }
                    }
                }
```

File: src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java [hunk 6]
```diff
@@ -510,4 +511,89 @@ public void testResultSetMethods() throws SQLException {
...
+                                assert (firstColumnValue.equalsIgnoreCase("No rowguidcol column defined."));
+                                break;
+                            case 5:
+                                assert (firstColumnValue.equalsIgnoreCase("PRIMARY"));
+                                break;
+                        }
+                    }
+                }
+
+                SQLWarning warnings = stmt.getWarnings();
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: Shouldn't we also validate the warnings?

Comment (cheenamalhotra) [reply src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: These warnings are not consistent and I don't think there is a clean way of testing them. Sometimes they're just blank lines. We just care about letting them pass through.

Comment (ulvii) [reply src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: Can we at least validate that warnings exist (not `null`)?

Comment (cheenamalhotra) [reply src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: Added a validation for their count (hardcoded value is based on what database responds with). We can confirm with this that all the warnings were captured.

**Pull request merged**